### PR TITLE
TP03

### PR DESCRIPTION
**Pull request by Cristiana959**
### Pull Request Overview

This pull request adds TP03.

### Testing Strategy

N/A

### TODO or Help Wanted
N/A

### Build

- [ ] Ran `npm build`.

### Author

Signed-off-by: Cristiana Andrei cristiana.andrei@wyliodrin.com

### CODE REVIEW

File: docs/tp/03.md [hunk 1]
```diff
@@ -0,0 +1,208 @@
+---
+sidebar_position: 2
```
Comment (alexandruradovici) [inline docs/tp/03.md:2]:
```suggestion
sidebar_position: 3
```

File: docs/tp/03.md [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+description: Une introduction au langage Rust.
```
Comment (alexandruradovici) [inline docs/tp/03.md:3]: The description is not the right one.

File: docs/tp/03.md [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+---
+
+# TP 02 - Systems de fichiers et le type Result
```
Comment (alexandruradovici) [inline docs/tp/03.md:6]:
```suggestion
# TP 03 - Systems de fichiers et le type Result
```

File: docs/tp/03.md [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+
+## Resources
+ 1. [Recoverable Errors with Result](https://jimskapt.github.io/rust-book-fr/ch09-02-recoverable-errors-with-result.html)
```
Comment (alexandruradovici) [inline docs/tp/03.md:9]:
```suggestion
 1. [Des erreurs récupérables avec Result](https://jimskapt.github.io/rust-book-fr/ch09-02-recoverable-errors-with-result.html)
```

File: docs/tp/03.md [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+ 9. [Linux permissions](https://en.wikipedia.org/wiki/Chmod)
+
+## Travail avec les fichiers en Rust
+Pour travailler avec des fichiers on doit utiliser le module fs qui se trouve dans la bilbiotheque standard de Rust.
+
+```rust
+use std::fs;
+```
+### Le struct File
+Le struct *File* renvoie un objet donnant accès à un fichier ouvert sur le système de fichiers.
```
Comment (alexandruradovici) [inline docs/tp/03.md:26]:
```suggestion
Le struct `File` renvoie un objet donnant accès à un fichier ouvert sur le système de fichiers.
```

File: docs/tp/03.md [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+
+Une instance d'un fichier peut être lue et/ou écrite en fonction des **options* avec lesquelles elle a été ouverte. Les fichiers implémentent également Seek pour modifier le curseur logique que le fichier contient en interne.
+
+Les fichiers sont **automatiquement fermés** lorsqu'ils sortent de **scope**. Les erreurs détectées à la fermeture sont ignorées par l'implémentation de Drop. Utilisez la méthode `sync_all` si ces erreurs doivent être gérées manuellement.
```
Comment (alexandruradovici) [inline docs/tp/03.md:30]:
```suggestion
Les fichiers sont **automatiquement fermés** lorsqu'ils sortent de **scope**. Les erreurs détectées à la fermeture sont ignorées par l'implémentation de `Drop`. Utilisez la méthode `sync_all` si ces erreurs doivent être gérées manuellement.
```

**Review by alexandruradovici — CHANGES_REQUESTED** [review]

**Review by alexandruradovici — APPROVED** [review]

File: docs/tp/03.md [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+Lorsqu'un programme demande d'ouvrir un fichier **le kernel** fait les actions suivantes:
+1. Accorde l'accès
+2. Crée une entrée dans la table de fichiers globale
+3. Fournit au logiciel l'emplacement de cette entrée
+
+Le descripteur est identifié par un **entier non négatif unique**, tel que 0, 12 ou 567. Au moins un descripteur de fichier existe pour chaque fichier ouvert sur le système.
+
+Pour obtenir le file descriptor d'un fichier on utilise la fonction [as_raw_fd()](https://doc.rust-lang.org/stable/std/os/fd/trait.AsRawFd.html#tymethod.as_raw_fd).
+
+## Le type Result
```
Comment (alexandruradovici) [inline docs/tp/03.md:98]: I would ,ove this before the functions for files, as those functions use it.

File: docs/tp/03.md [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+    println!("With text:\n{contents}");
+}
+```
+
+### Le struct Metadata
+Cette struct renvoie informations de métadonnées sur un fichier.
+
+Ce module contient des méthodes pour récupérer des informations sur un fichier telles que le type de fichier, les autorisations, etc.
+
+:::note
```
Comment (alexandruradovici) [inline docs/tp/03.md:82]:
```suggestion
:::tip
```

**Review by alexandruradovici — CHANGES_REQUESTED** [review]

File: docs/tp/03.md [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+```
+:::
+
+## Exercises 
+1. Écrivez un programme qui reçoit en paramètre un dossier. Imprimer le contenu du dossier.
+:::tip
+Utilisez [read_dir](https://doc.rust-lang.org/std/fs/fn.read_dir.html).
+:::
+
+2. Écrivez un programme qui reçoit en argument de ligne de commande un fichier et une tâche. Mettre en œuvre les tâches :
```
Comment (alexandruradovici) [inline docs/tp/03.md:201]:
```suggestion
2. Écrivez un programme qui reçoit en argument de ligne de commande un fichier et une tâche. Mettre en œuvre les tâches:
```

File: docs/tp/03.md [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+    a. **print** - imprime le contenu du fichier et le file descriptor
+    b. **size** - imprime la taille du fichier'
+    c. **owner** - affiche l'utilisateur (uid) et le groupe (gid) du fichier
+    d. **mode_number** - affiche le mode (permissions) en numero du fichier
+    e. **type** - le type de fichier: dossier, fichier, lien, périphérique caractère, périphérique bloc, …
+    f. **mode_text** - affiche le mode (permissions) en text (rwx…) du fichier
+    g. **print_buffer** - imprimez le contenu du fichier à l'aide des fonctions open et read.
```
Comment (alexandruradovici) [inline docs/tp/03.md:208]:
```suggestion
    1. **print** - imprime le contenu du fichier et le file descriptor
    2. **size** - imprime la taille du fichier'
    3. **owner** - affiche l'utilisateur (uid) et le groupe (gid) du fichier
    4. **mode_number** - affiche le mode (permissions) en numero du fichier
    5. **type** - le type de fichier: dossier, fichier, lien, périphérique caractère, périphérique bloc, …
    6. **mode_text** - affiche le mode (permissions) en text (rwx…) du fichier
    7. **print_buffer** - imprimez le contenu du fichier à l'aide des fonctions open et read.
```

**Review by alexandruradovici — CHANGES_REQUESTED** [review]

File: docs/tp/03.md (hunk 7 above)
Comment (Cristiana959) [reply docs/tp/03.md:98]: Done

Comment (alexandruradovici) [reply docs/tp/03.md:98]: We still have `Result` for the main function, maybe move it ahead of that?

**Review by alexandruradovici — CHANGES_REQUESTED** [review]

File: docs/tp/03.md [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+ 9. [Linux permissions](https://en.wikipedia.org/wiki/Chmod)
+
+## Travail avec les fichiers en Rust
+Pour travailler avec des fichiers on doit utiliser le module fs qui se trouve dans la bilbiotheque standard de Rust.
+
+```rust
+use std::fs;
+```
+
+## Le type Result
```
Comment (alexandruradovici) [inline docs/tp/03.md:26]: Place before *Travail avec les fichiers en Rust*

**Review by alexandruradovici — CHANGES_REQUESTED** [review]

**Review by alexandruradovici — APPROVED** [review]

**Pull request merged**